File: market-domain/src/main/java/com/example/market/domain/catalog/ProductId.java
package com.example.market.domain.catalog;

import java.util.Objects;
import java.util.UUID;

public record ProductId(UUID value) {
    public ProductId { Objects.requireNonNull(value); }
    public static ProductId newId() { return new ProductId(UUID.randomUUID()); }
    public static ProductId of(String s) { return new ProductId(UUID.fromString(s)); }
    @Override public String toString() { return value.toString(); }
}

File: market-domain/src/main/kotlin/com/example/market/domain/catalog/ProductId.kt
package com.example.market.domain.catalog

import java.util.UUID

/**
 * 상품 마스터 식별자.
 *
 * Java 호환을 유지하기 위해 일반 data class 로 둔다. value class 로 가면 Java 호출자에서
 * 매개변수 mangling 이 발생해 `findById(ProductId id)` 같은 메서드가 호출 불가능해진다.
 *
 * `@get:JvmName("value")` 로 기존 record accessor `value()` 를 그대로 보존.
 */
data class ProductId(@get:JvmName("value") val value: UUID) {

    override fun toString(): String = value.toString()

    companion object {
        @JvmStatic
        fun newId(): ProductId = ProductId(UUID.randomUUID())

        @JvmStatic
        fun of(s: String): ProductId = ProductId(UUID.fromString(s))
    }
}
